burn-tensor/src/tensor/api/base.rs

```diff
@@ -67,6 +67,60 @@ where
         Tensor::new(K::reshape::<D, D2>(self.primitive, shape))
     }
 
+    /// Reshape the tensor to have the given shape.
+    ///
+    /// Compared to [Tensor::reshape](Tensor::reshape),
+    /// this function allows to infer the size of one dimension and matches the numpy/pytorch behavior.
+    ///
+    /// # Arguments
+    ///
+    /// - `shape`: The new shape of the tensor.
+    ///
+    /// # Panics
+    ///
+    /// - If the tensor contains more than one `-1` in the shape.
+    /// - If the tensor contains values that are not positive (other than -1).
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use burn_tensor::backend::Backend;
+    /// use burn_tensor::Tensor;
+    ///
+    /// fn example<B: Backend>() {
+    ///    let tensor = Tensor::<B, 3>::ones([2, 3, 4]);
+    ///
+    ///    // Given a 3D tensor with dimensions (2, 3, 4), reshape it to (2, 12)
+    ///    let reshaped_tensor: Tensor::<B, 2> = tensor.reshape_infer([2, -1]);
+    ///
+    ///    // The resulting tensor will have dimensions (2, 12).
+    ///    println!("{:?}", reshaped_tensor.shape());
+    ///
+    /// }
+    ///
+    pub fn reshape_infer<const D2: usize>(self, shape: [i64; D2]) -> Tensor<B, D2, K> {
```

Comment:
I don't think we should have another reshape method, but I agree that it would be convenient to support negative numbers. So, I propose to implement `From<[i64; D]> for Shape<D>` instead of adding a new function. This will have the benefit of a more uniform API.

Maybe it will clash with the current `From<[usize; D]> for Shape<D>` implementation, where users are forced to specify the array type (bad API). In that case, I propose to only keep the `From<i64>`.

Reply:
I agree your suggestion is better. I had introduced reshape_infer to avoid compatibility with the existing usage but we found a way to accomplish without introducing any disruption. 

The interface is combined now into one method `reshape`. With your help, I introduced ReshapeArgs that serves as intermediate type that helps conversion from `[i32;D]`, `[usize;D]`, and Shape.